include/experimental/__p2630_bits/submdspan_mapping.hpp

```diff
@@ -237,124 +334,141 @@ struct preserve_layout_right_mapping<std::index_sequence<Idx...>, SubRank,
+      detail::any_slice_out_of_bounds(this->extents(), slices...);
   auto offset = static_cast<size_t>(
-    out_of_bounds ?
-    this->required_span_size() :
-    this->operator()(detail::first_of(slices)...)
-  );
-  
+      out_of_bounds ? this->required_span_size()
+                    : this->operator()(detail::first_of(slices)...));
+
   if constexpr (std::is_same_v<dst_layout_t, layout_right>) {
     // layout_right case
-    return submdspan_mapping_result<dst_mapping_t>{dst_mapping_t(dst_ext), offset};
-  } else {
-    // layout_stride case
-    auto inv_map = detail::inv_map_rank(
-      std::integral_constant<size_t,0>(),
-      std::index_sequence<>(),
-      slices...);
+    return submdspan_mapping_result<dst_mapping_t>{dst_mapping_t(dst_ext),
+                                                   offset};
+  } else if constexpr (std::is_same_v<
+                           dst_layout_t,
+                           MDSPAN_IMPL_PROPOSED_NAMESPACE::layout_right_padded<
+                               dynamic_extent>>) {
     return submdspan_mapping_result<dst_mapping_t>{
-        dst_mapping_t(dst_ext, detail::construct_sub_strides(
-                                   *this, inv_map,
-    // HIP needs deduction guides to have markups so we need to be explicit
-    // NVCC 11.0 has a bug with deduction guide here, tested that 11.2 does not have the issue
-    // But Clang-CUDA also doesn't accept the use of deduction guide so disable it for CUDA alltogether
-    #if defined(_MDSPAN_HAS_HIP) || defined(_MDSPAN_HAS_CUDA)
-                                   std::tuple<decltype(detail::stride_of(slices))...>{detail::stride_of(slices)...})),
-    #else
-                                   std::tuple{detail::stride_of(slices)...})),
-    #endif
+        dst_mapping_t(dst_ext,
+                      stride(src_ext_t::rank() - 2 - deduce_layout::NumGaps)),
         offset};
+  } else {
+    // layout_stride case
+    auto inv_map = detail::inv_map_rank(std::integral_constant<size_t, 0>(),
+                                        std::index_sequence<>(), slices...);
+    return submdspan_mapping_result<dst_mapping_t> {
+      dst_mapping_t(dst_ext,
+                    detail::construct_sub_strides(
+                        *this, inv_map,
+// HIP needs deduction guides to have markups so we need to be explicit
+// NVCC 11.0 has a bug with deduction guide here, tested that 11.2 does not have
+// the issue But Clang-CUDA also doesn't accept the use of deduction guide so
+// disable it for CUDA alltogether
+#if defined(_MDSPAN_HAS_HIP) || defined(_MDSPAN_HAS_CUDA)
+                        std::tuple<decltype(detail::stride_of(slices))...>{
+                            detail::stride_of(slices)...})),
+#else
+                        std::tuple{detail::stride_of(slices)...})),
+#endif
+          offset
+    };
   }
 #if defined(__NVCC__) && !defined(__CUDA_ARCH__) && defined(__GNUC__)
   __builtin_unreachable();
```

Comment:
Tangent: This GCC built-in works fine for us in device code (with `__CUDA_ARCH__` defined).